src/mrpro/algorithms/optimizers/pgd.py

```diff
@@ -0,0 +1,158 @@
+"""Proximal Gradient Descent algorithm."""
+
+import math
+from collections.abc import Callable
+
+import torch
+from typing_extensions import Unpack
+
+from mrpro.algorithms.optimizers.OptimizerStatus import OptimizerStatus
+from mrpro.operators import ProximableFunctionalSeparableSum
+from mrpro.operators.Functional import ProximableFunctional
+from mrpro.operators.Operator import Operator
+
+
+class PGDStatus(OptimizerStatus):
+    """Proximal Gradient Descent callback base class."""
+
+    stepsize: float
+    """Current stepsize of the algorithm."""
+
+
+def pgd(
+    f: Operator[torch.Tensor, tuple[torch.Tensor]] | Operator[Unpack[tuple[torch.Tensor, ...]], tuple[torch.Tensor]],
+    g: ProximableFunctional | ProximableFunctionalSeparableSum,
+    initial_value: torch.Tensor | tuple[torch.Tensor, ...],
+    stepsize: float = 1.0,
+    reg_parameter: float = 0.01,
+    max_iterations: int = 128,
+    backtrack_factor: float = 1.0,
+    callback: Callable[[PGDStatus], None] | None = None,
+) -> tuple[torch.Tensor, ...]:
+    r"""Proximal gradient descent algorithm for solving problem :math:`min_x f(x) + g(x)`.
+
+    f is convex, differentiable, and with L-Lispchitz gradient.
+    g is convex, non-smooth with computable proximal map.
+
+    For fixed stepsize t, pgd converges globally when :math:`t \in (0, 1/L)`,
+    where L is the Lipschitz constant of the gradient of f.
+    In applications, f is usually of the form :math:`f(x) = 1/2 ||Ax - target||^2`,
+    where :math:`A` is a linear operator.
+    In this case, :math:`t \in (0, 1/||A^T A||)` for convergence.
+    If no backtracking is used, the fixed stepsize should be given accordingly to the convergence condition.
+
+    Example:
+        L1 regularized image reconstruction. Problem formulation: :math:`min_x 1/2 ||Fx - target||^2 + ||x||_1`,
```

Comment:
good idea to have an example here. At least, I like it. :)  Should we have it for pdhg as well, @fzimmermann89?

Reply:
Good idea. I will also try out the doctest pytest plugin, to check that this example code actually runs (and is not forgotten on changes to the code)